### DISCUSSION TITLE
long export time when the db is big

### DISCUSSION THREAD

**Discussion by aniel300** [sonroyaalmerol/m3u-stream-merger-proxy#138] [General]
this is in reference to https://github.com/sonroyaalmerol/m3u-stream-merger-proxy/issues/72#issuecomment-2299939900 I was trying to capture an interesting line of the log that day but i could not do it, but i think i just bumped into by mistake the line is this `2024/08/23 22:57:17 [DEBUG] Cache miss. Retrieving streams from Redis...` not sure if this will be helpful or not.

**Comment by aniel300** [comment]
here is some more logs. it looks like it is suck however i just confirmed the two groups are present in the db/m3u so we good there. now the only problem is when the the db is big
![image](https://github.com/user-attachments/assets/b7dbd541-d1f5-4e1e-b9c4-8a0f2135f032)

**Comment by aniel300** [comment]
please let me know if am leaking info in this screenshot and if u can a way to prevent info leak when using the debug will be awesome because manually making sure and removing the URL is a pain.

**Comment by sonroyaalmerol** [comment]
Can you test and see if #124 results to faster M3U returns? I've reworked the caching behind the scenes. It still should have a similar behavior where the cache needs to be built completely on first request. Also `CACHE_ON_SYNC` will require `BASE_URL` to be set as well.

`BASE_URL` sets the base URL for the stream URLs in the M3U file to be generated (e.g. <http://192.168.1.10:8080>). This data is usually gathered from the HTTP requests done by the client. However, `CACHE_ON_SYNC` does its thing in the background without any HTTP requests so it needs to be given the base URL manually.

**Comment by aniel300** [comment]
base url is where the proxy server lives?

**Comment by sonroyaalmerol** [comment]
Yep! Not the base URL of the source streams.

**Comment by aniel300** [comment]
ok let me test and report back

**Comment by aniel300** [comment]
every time u mentions a specific pr does it means it is also available in dev tag as well?

**Comment by sonroyaalmerol** [comment]
Only if the PR has been merged. There are specific PRs where I try not to merge them immediately as it changes a lot of components.

**Comment by aniel300** [comment]
i have tried all possible combinations such as `http://172.18.0.1:8086` and `http://172.18.0.1:8080` and still getting `2024/08/25 18:55:43 [DEBUG] Cache miss. Retrieving streams from Redis...` in the logs how do i know the new implementation is taking effect? also `SAFE_LOGS` goes as `SAFE_LOGS=true`?

**Comment by sonroyaalmerol** [comment]
Well, the fact that you're seeing that log means that you're not using the image of the PR I mentioned. PRs that I mention will have a comment from a bot containing the image URL for that specific PR.

For #124, it should be this [comment](https://github.com/sonroyaalmerol/m3u-stream-merger-proxy/pull/124#issuecomment-2308987426). Use that image URL instead of the usual `sonroyaalmerol/m3u-stream-merger-proxy:dev` instead to test a specific PR.

`SAFE_LOGS=true` is correct. However, the base urls you provided doesn't seem like the IP addresses I would expect in a local setup.

To make things simpler for you, the base URL can be derived from the URL you use to access the generated M3U.

For example, if you access the M3U with this URL: `http://192.168.2.5:8080/playlist.m3u`, then the base URL would be `http://192.168.2.5:8080`. You use that as the value for `BASE_URL` when trying out the PR image.

**Comment by aniel300** [comment]
rest assured am trying the correct image and those ip are from the docker network on my debain 12 linux machine. i will share screenshot of everything once i get back home, thank you.

**Comment by sonroyaalmerol** [comment]
Sure. Do double check as it is actually impossible for that PR image to return that log line. It doesn't exist in the code of that PR.

**Comment by aniel300** [comment]
current image being use and docker ip/network.
![image](https://github.com/user-attachments/assets/6f4b882f-7a8d-4e5b-b313-ef30134c5fd5)
![image](https://github.com/user-attachments/assets/acfc11a7-c8fa-4176-9f38-9219323dee8e)
![image](https://github.com/user-attachments/assets/7ef2615d-13a0-4c03-80ff-2a0e024831e2)

**Comment by sonroyaalmerol** [comment]
I guess it's the workflow not doing what it's supposed not building the right image anymore in PRs for some reason. :facepalm: I'll merge it to dev. You can test it from there instead.

**Comment by aniel300** [comment]
ok thanks

**Comment by aniel300** [comment]
unfortunately, the issue has not been fixed. here is some screenshot of what it has been doping. db is around 300mb btw. i left it all night because i had to go to sleep.
![msedge_W8oaqPtISz](https://github.com/user-attachments/assets/08d26005-c73f-4cb1-962f-2d84f27da5f4)
![msedge_0pcR1aO8XW](https://github.com/user-attachments/assets/456f001d-2b3a-42ff-9de5-ce71ab15c5b6)

**Comment by sonroyaalmerol** [comment]
Can you give me more context: what makes you say it doesn't work? What is the output when you request for the M3U url?

**Comment by aniel300** [comment]
it keeps loading in the browser or if i try to download the m3u via idm it never downloads. there is a little of cpu spike on the db container and proxy container (I believe) but that is about it. for example, the small channel sample am sing for debugging (~38 channels) works as expected, sync is quick and same goes for export but when u give it multiple url/providers that comes with vod, etc which all together is around 300mb in size then export doesn't seem to work. i dont the sync taking time the firs t time but the exports after that i would like to be quick.

**Comment by sonroyaalmerol** [comment]
I did more work for this issue. I've merged the changes to the dev build if you want to try it out again.

**Comment by aniel300** [comment]
is there any way to make the db use 100 percent of the cpu so maybe the whole process can be speed up as it is doing a lot of this `2024/08/26 19:58:13 [DEBUG] Processing stream: xxx map[]}` ? 
![image](https://github.com/user-attachments/assets/7a28b6d1-ed50-444c-907d-c70b9fdcaac4)

**Comment by aniel300** [comment]
could the slowness be caused by the sorting? it still doing a lot of this and there is no progress bar so idk how much time is left for this to finish 
![image](https://github.com/user-attachments/assets/0a1fbf1d-e19c-42de-b220-e7404e94e8d5)

**Comment by sonroyaalmerol** [comment]
Please be patient and wait for everything to be processed to cache. We're talking about ~300mb worth of strings being processed. How I wish it is as easy as simply setting the "amount of CPU" to be used. That is just not how software works. Most of the process required for the proxy to work are single-threaded and cannot be parallelized.

Also, do disable debugging mode if you want the maximum performance possible. Logging affects performance more than you might expect especially in cases like these. Logging is a single-threaded process which pretty much forces even the parallelized processes to wait for the log to be printed to the terminal before the next job is executed.

The sorting is index-based and does not add any time complexity at all. Once everything is processed, the M3U will be stored in plain-text, both in memory and as a file. At that point, the only bottleneck would be your RAM speed and disk i/o.

The function for time complexity will always be directly proportional to the amount of ingress data.

**Comment by sonroyaalmerol** [comment]
I've just merged to dev more optimizations. This will probably be something that will improve over time as I fix other issues as a side effect. I won't be focusing on this anymore in the near future. Converting this to a discussion instead.

**Comment by aniel300** [comment]
it didn't work in the end even after disabling debugging, and yes i agree with you and i thank u for working and trying so hard on this. i will try again later on with no vod and only channels to see if that makes a difference.
![IDMan_8Luk2I6EsH](https://github.com/user-attachments/assets/bdf02659-b0b3-4fbe-9822-8956d8a4f041)

**Comment by sonroyaalmerol** [comment]
@aniel300 I've done some major refactoring and optimization with #144 which will decrease the number of queries the proxy does with the database by at least half. I'm not sure how that would translate in an actual set of streams. Test it out in the dev build with `CLEAR_ON_BOOT` set to true and `DEBUG` set to false. Let me know how it goes.

**Reply by aniel300** [reply]
will do. am also eager to test with vod out of the equation to see the size of the db and see if that makes a difference, thank you for continue giving this a try. if this happens to be a vod/ db being to big i can settle for channels only 😉🙏

**Reply by aniel300** [reply]
getting some go crashes i think and db still at 1kb. not sure if am supposed to keep waiting

sync.runtime_SemacquireMutex(0x0?, 0x0?, 0xc0008024d0?)
	/usr/local/go/src/runtime/sema.go:95 +0x25
sync.(*Mutex).lockSlow(0xc00012c018)
	/usr/local/go/src/sync/mutex.go:173 +0x15d
sync.(*Mutex).Lock(...)
	/usr/local/go/src/sync/mutex.go:92
m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:225 +0x22d
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 14
	/app/m3u/parser.go:219 +0x65b
goroutine 115 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:221 +0xc5
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 14
	/app/m3u/parser.go:219 +0x65b
goroutine 117 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:221 +0xc5
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 14
	/app/m3u/parser.go:219 +0x65b
goroutine 118 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:221 +0xc5
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 14
	/app/m3u/parser.go:219 +0x65b
goroutine 119 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:221 +0xc5
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 14
	/app/m3u/parser.go:219 +0x65b
goroutine 120 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:221 +0xc5
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 14
	/app/m3u/parser.go:219 +0x65b
goroutine 121 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func2()
	/app/m3u/parser.go:238 +0x6b
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 14
	/app/m3u/parser.go:237 +0x769
goroutine 122 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:221 +0xc5
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 13
	/app/m3u/parser.go:219 +0x65b
goroutine 123 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:221 +0xc5
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 13
	/app/m3u/parser.go:219 +0x65b
goroutine 124 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:221 +0xc5
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 13
	/app/m3u/parser.go:219 +0x65b
goroutine 125 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:221 +0xc5
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 13
	/app/m3u/parser.go:219 +0x65b
goroutine 126 [runnable]:
m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:221 +0xc5
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 13
	/app/m3u/parser.go:219 +0x65b
goroutine 127 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:221 +0xc5
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 13
	/app/m3u/parser.go:219 +0x65b
goroutine 128 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:221 +0xc5
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 13
	/app/m3u/parser.go:219 +0x65b
goroutine 129 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:221 +0xc5
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 13
	/app/m3u/parser.go:219 +0x65b
goroutine 130 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func2()
	/app/m3u/parser.go:238 +0x6b
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 13
	/app/m3u/parser.go:237 +0x769

**Reply by aniel300** [reply]
with CLEAR_ON_BOOT off am getting this logs

m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:221 +0xc5
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 37
	/app/m3u/parser.go:219 +0x65b
goroutine 119 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:221 +0xc5
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 37
	/app/m3u/parser.go:219 +0x65b
goroutine 120 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:221 +0xc5
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 37
	/app/m3u/parser.go:219 +0x65b
goroutine 121 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:221 +0xc5
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 37
	/app/m3u/parser.go:219 +0x65b
goroutine 122 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:221 +0xc5
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 37
	/app/m3u/parser.go:219 +0x65b
goroutine 123 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:221 +0xc5
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 37
	/app/m3u/parser.go:219 +0x65b
goroutine 124 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:221 +0xc5
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 37
	/app/m3u/parser.go:219 +0x65b
goroutine 125 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:221 +0xc5
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 37
	/app/m3u/parser.go:219 +0x65b
goroutine 126 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func2()
	/app/m3u/parser.go:238 +0x6b
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 37
	/app/m3u/parser.go:237 +0x769
goroutine 127 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:221 +0xc5
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 36
	/app/m3u/parser.go:219 +0x65b
goroutine 128 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:221 +0xc5
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 36
	/app/m3u/parser.go:219 +0x65b
goroutine 129 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:221 +0xc5
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 36
	/app/m3u/parser.go:219 +0x65b
goroutine 130 [runnable]:
m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:221 +0xc5
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 36
	/app/m3u/parser.go:219 +0x65b
goroutine 131 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:221 +0xc5
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 36
	/app/m3u/parser.go:219 +0x65b
goroutine 132 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:221 +0xc5
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 36
	/app/m3u/parser.go:219 +0x65b
goroutine 134 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func1()
	/app/m3u/parser.go:221 +0xc5
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 36
	/app/m3u/parser.go:219 +0x65b
goroutine 135 [chan receive]:
m3u-stream-merger/m3u.ParseM3UFromURL.func2()
	/app/m3u/parser.go:238 +0x6b
created by m3u-stream-merger/m3u.ParseM3UFromURL in goroutine 36
	/app/m3u/parser.go:237 +0x769

**Reply by aniel300** [reply]
i just notice that with `with CLEAR_ON_BOOT off` it is moving along despite above but idk

**Comment by aniel300** [comment]
where can i get this image? https://github.com/sonroyaalmerol/m3u-stream-merger-proxy/pull/143 as latest dev broke something form me. i will make a new container tomorrow for testing and debugging that way i don't have to be going back fort.
